src/decoder/lossless.rs

```diff
@@ -0,0 +1,212 @@
+                        if mcu_x > 0 {
+                            rc[i] = results[i]
+                                [(mcu_y - 1) * frame.image_size.width as usize + (mcu_x - 1)];
+                        }
+                    }
+                    let prediction = predict(
+                        ra[i] as i32,
+                        rb[i] as i32,
+                        rc[i] as i32,
+                        scan.predictor_selection,
+                        scan.point_transform,
+                        frame.precision,
+                        mcu_x,
+                        mcu_y,
+                        self.restart_interval > 0
+                            && mcus_left_until_restart == self.restart_interval - 1,
+                    );
+                    let result = ((prediction + diff) & 0xFFFF) as u16; // modulo 2^16
+                    results[i].push(result << scan.point_transform);
+                }
+            }
+        }
+
+        let mut marker = huffman.take_marker(&mut self.reader)?;
+        while let Some(Marker::RST(_)) = marker {
+            marker = self.read_marker().ok();
+        }
+        Ok((marker, Some(results)))
+    }
+}
+
+/// H.1.2.1
+fn predict(
+    ra: i32,
+    rb: i32,
+    rc: i32,
+    predictor: Predictor,
+    point_transform: u8,
+    input_precision: u8,
+    ix: usize,
+    iy: usize,
+    restart: bool,
+) -> i32 {
+    let result = if (ix == 0 && iy == 0) || restart {
+        // start of first line or restart
+        1 << (input_precision - point_transform - 1)
+    } else if iy == 0 {
+        // rest of first line
+        ra
+    } else if ix == 0 {
+        // start of other line
+        rb
+    } else {
+        // use predictor Table H.1
+        match predictor {
+            Predictor::NoPrediction => 0,
+            Predictor::Ra => ra,
+            Predictor::Rb => rb,
+            Predictor::Rc => rc,
+            Predictor::RaRbRc1 => ra + rb - rc,
+            Predictor::RaRbRc2 => ra + ((rb - rc) >> 1),
+            Predictor::RaRbRc3 => rb + ((ra - rc) >> 1),
+            Predictor::RaRb => (ra + rb) / 2,
+        }
+    };
+    result
+}
+
+pub fn compute_image_lossless(
+    frame: &FrameInfo,
+    mut data: Vec<Vec<u16>>
+) -> Result<Vec<u8>> {
+    if data.is_empty() || data.iter().any(Vec::is_empty) {
+        return Err(Error::Format("not all components have data".to_owned()));
+    }
+    let output_size = frame.output_size;
+    let components = &frame.components;
+    let ncomp = components.len();
+
+    if ncomp == 1 {
+        let decoded = convert_to_u8(frame, data.remove(0));
+        Ok(decoded)
+    } else {
+        let mut decoded: Vec<u16> = vec![
+            0u16;
+            ncomp * output_size.width as usize * output_size.height as usize
+        ];
+        for (i, component) in components.iter().enumerate() {
+            for x in 0 .. data[i].len() {
+                decoded[x * ncomp + i] = data[i][x];
+            }
+        }
```

Comment:
Also slower than necessary. We're passing over `decoded` slice a total of `ncomp` times, each time writing single bytes into larger chunks. This should be the other way around:

```suggestion
        for (x, chunk) in decoded.chunks_mut(ncomp).enumerate() {
            for (i, (component_data, _)) in data.iter().zip(&components).enumerate() {
                chunk[i] = component_data[x];
            }
        }
```

Reply:
thanks, that works